### PR TITLE
[JobLauncher] modify webhook

### LINKED ISSUE #1506: Job Launcher - Modify webhook endpoint
**Description**
The endpoint `/escrow-failed-webhook` should be a POST to `/webhook` in webhook controller.
It should follow this standard https://tech-docs.humanprotocol.org/architecture/exchange-oracle#webhook

**Motivation**
Follow defined standard for webhooks

**Implementation details (To be defined by developer)**

### PR DESCRIPTION
**Pull request by mrhouzlane**
## Description

Follows https://tech-docs.humanprotocol.org/architecture/exchange-oracle#webhook for Job Launcher. 

## Summary of changes

- Add `webhookController` for the Job Launcher 

Closes #1506

### CODE REVIEW

File: packages/apps/job-launcher/server/src/common/enums/webhook.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 export enum EventType {
   ESCROW_CREATED = 'escrow_created',
   ESCROW_CANCELED = 'escrow_canceled',
-  TASK_CREATION_FAILED = 'task_creation_failed',
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/common/enums/webhook.ts:4]: The event type for this case is not submissions rejected, it is escrow_failed

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.controller.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  })
+  @ApiResponse({
+    status: 400,
+    description: 'Bad Request. Invalid input parameters.',
+  })
+  @ApiResponse({
+    status: 401,
+    description: 'Unauthorized. Missing or invalid credentials.',
+  })
+  @Post('/escrow-failed-webhook')
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/modules/webhook/webhook.controller.ts:34]: the route following the docs should be` /webhook`

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts [hunk 3]
```diff
@@ -1,8 +1,24 @@
...
+import {
+  IsEnum,
+  IsObject,
+  IsString,
+  IsBoolean,
+  IsOptional,
+} from 'class-validator';
 import { EventType, OracleType } from '../../common/enums/webhook';
 
+export class EventData {
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts:12]: In this case event data should be something like:
```
{
  "reason": "Invalid manifest" (This reason is just an example)
}
```

**Review by portuu3 — CHANGES_REQUESTED** [review]

File: packages/apps/job-launcher/server/src/common/enums/webhook.ts (hunk 1 above)
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/common/enums/webhook.ts:4]: Changed

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.controller.ts (hunk 2 above)
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/modules/webhook/webhook.controller.ts]: Changed

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts (hunk 3 above)
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts]: Added Invalid Manifest and Manifest Cannot Be Downloaded as 2 possible reasons

File: packages/apps/job-launcher/server/src/common/validators/ethers.ts
Comment (flopez7) [inline packages/apps/job-launcher/server/src/common/validators/ethers.ts:1]: You don't need that, there is already a validator called `@IsEthereumAddress()` inside `class-validator`

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.controller.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  @ApiResponse({
+    status: 400,
+    description: 'Bad Request. Invalid input parameters.',
+  })
+  @ApiResponse({
+    status: 401,
+    description: 'Unauthorized. Missing or invalid credentials.',
+  })
+  @Post()
+  public async handleEscrowFailedWebhook(
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/webhook/webhook.controller.ts:35]: The endpoint should be generic for all type of events, not only `escrow_failed`
Please take a look here for it: https://github.com/humanprotocol/human-protocol/tree/develop/packages/apps/fortune/exchange-oracle/server/src/modules/webhook

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts [hunk 5]
```diff
@@ -1,8 +1,37 @@
...
+  @ApiProperty({ name: 'manifest_cannot_be_downloaded' })
+  @IsString()
+  manifestCannotBeDownloaded: string;
+
+  @ApiProperty()
+  @IsString()
+  reason?: string;
+}
+
+export type EventData = InvalidManifest | ManifestCannotBeDownloaded;
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts:33]: This is not needed. You can use a simple object for both types of evets like:
```suggestion
export class EventData {
  @ApiProperty()
  @IsString()
  reason?: string;
}
```

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts [hunk 6]
```diff
@@ -31,12 +60,18 @@ export class WebhookDataDto {
   public chainId: ChainId;
 
   @ApiProperty({ name: 'escrow_address' })
+  @IsString()
+  @IsValidEthereumAddress()
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts:64]: use the validator from `class-validator`

**Review by flopez7 — CHANGES_REQUESTED** [review]

File: packages/apps/job-launcher/server/src/common/validators/ethers.ts
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/common/validators/ethers.ts:1]: changed to this : `import { IsEthereumAddress } from 'class-validator';`

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts (hunk 6 above)
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts:64]: changed

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.controller.ts (hunk 4 above)
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/modules/webhook/webhook.controller.ts:35]: changed to handle webhook for the 3 event types

**Review by flopez7 — CHANGES_REQUESTED** [review]

File: packages/apps/job-launcher/server/src/modules/job/job.service.ts [hunk 7]
```diff
@@ -1016,6 +1016,31 @@ export class JobService {
...
+
+    if ('manifestCannotBeDownloaded' in dto.eventData) {
+      this.logger.log(
+        `Manifest cannot be downloaded: ${dto.eventData.manifestCannotBeDownloaded}`,
+        JobService.name,
+      );
+      throw new BadRequestException(
+        `Manifest cannot be downloaded: ${dto.eventData.manifestCannotBeDownloaded}`,
+      );
+    }
```
[the remark this answers is not in the document]
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/modules/job/job.service.ts]: removed

File: packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts [hunk 8]
```diff
@@ -1,8 +1,35 @@
...
+  })
+  @IsString()
+  @IsOptional()
+  manifestCannotBeDownloaded?: string;
+
+  @ApiProperty()
+  @IsString()
+  @IsOptional()
+  reason?: string;
+}
```
[the remark this answers is not in the document]
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/modules/webhook/webhook.dto.ts]: changed

File: packages/apps/job-launcher/server/src/common/enums/role.ts [hunk 9]
```diff
@@ -2,4 +2,5 @@ export enum Role {
   Exchange = 'exchange',
   Recording = 'recording',
   Reputation = 'reputation',
+  JobLauncher = 'job-launcher',
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/common/enums/role.ts:5]:
```suggestion
```

File: packages/apps/job-launcher/server/src/common/constants/index.ts [hunk 10]
```diff
@@ -23,7 +23,7 @@ export const MAINNET_CHAIN_IDS = [
 export const SENDGRID_API_KEY_REGEX =
   /^SG\.[A-Za-z0-9-_]{22}\.[A-Za-z0-9-_]{43}$/;
 
-export const HEADER_SIGNATURE_KEY = 'human-signature';
+export const HEADER_SIGNATURE_KEY = 'Human-Signature';
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/common/constants/index.ts:26]: We have it in lower case in all oracles, let's keep it 
```suggestion
export const HEADER_SIGNATURE_KEY = 'human-signature';
```

File: packages/apps/job-launcher/server/src/common/enums/webhook.ts [hunk 11]
```diff
@@ -1,7 +1,7 @@
...
+  ESCROW_FAILED = 'escrow_failed',
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/common/enums/webhook.ts:4]: Can both be added? Cvat oracles are working with `TASK_CREATION_FAILED` so we'll keep it until they move to the new `ESCROW_FAILED`
```suggestion
  TASK_CREATION_FAILED = 'task_creation_failed',
  ESCROW_FAILED = 'escrow_failed',
```

**Review by flopez7 — CHANGES_REQUESTED** [review]

File: packages/apps/job-launcher/server/src/common/enums/role.ts (hunk 9 above)
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/common/enums/role.ts]: removed

File: packages/apps/job-launcher/server/src/common/constants/index.ts (hunk 10 above)
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/common/constants/index.ts]: changed

File: packages/apps/job-launcher/server/src/common/enums/webhook.ts (hunk 11 above)
Comment (mrhouzlane) [reply packages/apps/job-launcher/server/src/common/enums/webhook.ts:5]: added

**Review by flopez7 — APPROVED** [review]

**Review by portuu3 — APPROVED** [review]

**Pull request merged**